src/lib/signature.ts

```diff
@@ -192,12 +198,23 @@ class Signature extends CircuitValue {
   static create(privKey: PrivateKey, msg: Field[]): Signature {
     const publicKey = PublicKey.fromPrivateKey(privKey).toGroup();
     const d = privKey.s;
-    const kPrime = Scalar.random();
+    const kPrime = Scalar.fromBigInt(
+      deriveNonce(
+        { fields: msg.map((f) => f.toBigInt()) },
+        { x: publicKey.x.toBigInt(), y: publicKey.y.toBigInt() },
+        BigInt(d.toJSON()),
+        'testnet'
+      )
+    );
     let { x: r, y: ry } = Group.generator.scale(kPrime);
     const k = ry.toBits()[0].toBoolean() ? kPrime.neg() : kPrime;
-    const e = Scalar.fromBits(
-      Poseidon.hash(msg.concat([publicKey.x, publicKey.y, r])).toBits()
+    let h = hashWithPrefix(
+      prefixes.signatureTestnet,
```

Comment:
I chose to hard-code the signature to 'testnet', since the distinction between testnet and mainnet for signatures which aren't about transactions seemed not useful to me and just adds noise IMO

Reply:
That's a good point! It opens a can of worms though. When signing transactions, the signature verifier is the Mina node, which has a well-defined network - ti's either a mainnet node or a testnet node, so it knows which one it should use for verifying.
However, a smart contract currently doesn't have that. Do you think the network type should be hard-coded into the smart contract / circuit somehow? Should we leave it as a constant for the developer to define? (But then we're pushing to them the problem of switching this flag when deploying / compiling to either network).

I think, if we go this route, the network type should definitely be a constant in the circuit and not a `Bool` that the developer can make variable. Because if it can be variable, it's too easy a mistake to just make it a private user input which doesn't constrain the proofs to be valid for the respective network.

Reply:
Hmm I suppose it’s much less dangerous than a transaction so I think it’s not worth the complexity cost of making these signatures distinct, but we should probably run this by more people. I’ll approve but I’m going to send a link to one of our slack channels and let’s let it bake overnight if anyone disagrees before we land it

Reply:
am I understanding this right that the constant doesn't matter here because it's what the zkapp will verify and so you could pretty much use anything?

Reply:
@mimoo yes, doesn't matter here in the sense that these signatures won't be sent to either network. we are completely in user-land: signatures are inputs to proofs created locally, where they are verified in-snark.

I think Brandon highlights an important problem: we don't provide replay protection for the user here, especially given that signatures are deterministic (nonce is derived from the message and private key). so, we currently leave it up to the application to figure out how to prevent that.

I think its too deep and broad a problem to solve in the context of this PR, but I created an issue so we don't forget to discuss this: https://github.com/o1-labs/snarkyjs/issues/720